pyparsing/exceptions.py

```diff
@@ -162,6 +162,23 @@ def column(self) -> int:
         """
         return col(self.loc, self.pstr)
 
+    @property
+    def found(self) -> str:
+        if not self.pstr:
+            return ""
+
+        if self.loc >= len(self.pstr):
+            return "end of text"
+
+        # pull out next word at error location
+        found_match = _exception_word_extractor.match(self.pstr, self.loc)
+        if found_match is not None:
+            found = found_match.group(0)
+        else:
+            found = self.pstr[self.loc : self.loc + 1]
+
+        return repr(found).replace(r"\\", "\\")
+
```

Comment:
This looks good, but I think I want to make some minor refactoring changes:
- change this and all the other properties to `cached_property`, imported from functools
- move the formatting that is done in `__str__` to a new method, `formatted_message` (see below)
- change `__str__` to just return `self.formatted_message()`

```python
    def formatted_message(self):
        found_phrase = f", found {self.found}" if self.found else ""
        return f"{self.msg}{found_phrase}  (at char {self.loc}), (line:{self.lineno}, col:{self.column})"
```

(I liked `found_phrase` over `foundstr`, as I advance in my pyparsing support years, I'm trying to make variable names more readable and intuitive.)

This will allow you to format your own custom exception messages, like this (inserted at the beginning of your program, to monkeypatch the default `formatted_message` function).

```python
from pyparsing import Keyword, ParseBaseException

def custom_exception_message(exc):
    found_phrase = f", found {exc.found}" if exc.found else ""
    return f"{exc.lineno}:{exc.column} {exc.msg}{found_phrase}"

ParseBaseException.formatted_message = custom_exception_message

Keyword("KEY1").parse_string("KEY2", parseAll=True)
```